### PR TITLE
support UDFs

### PR DESCRIPTION
**Pull request by pgrivachev**
Added support for persistent user-defined functions (inline is not supported).

### CODE REVIEW

File: server/src/test/ZetaSqlWrapper.spec.ts [hunk 1]
```diff
@@ -172,12 +176,38 @@ describe('ZetaSqlWrapper', () => {
...
+        typeKind: TypeKind.TYPE_INT64,
+      },
+    };
+
+    const rootCatalog = registerUdf(zetaSqlWrapper, udf);
+
+    const udfsCatalog = rootCatalog.catalog?.find(c => c.name === 'udfs');
+    assert.ok(udfsCatalog);
+    const func = udfsCatalog.customFunction?.find(c => c.namePath && arraysAreEqual(c.namePath, ['udfs', 'func']));
+    assert.ok(func);
```
Comment (fivetran-dmitryboykov) [inline server/src/test/ZetaSqlWrapper.spec.ts:211]: Do we need to check registered arguments, return type?

**Review by fivetran-dmitryboykov — CHANGES_REQUESTED** [review]
Some notes and 2 questions:
1. If we can extract udf’s from AST analyse result?
2. Do we need to remove udf’s from catalog?

File: server/src/ZetaSqlWrapper.ts [hunk 2]
```diff
@@ -118,6 +122,65 @@ export class ZetaSqlWrapper {
...
+    for (const namePath of namePaths) {
+      const udf = await this.createUdfFromNamePath(namePath);
+      if (udf) {
+        this.registerUdf(udf);
+      }
+    }
+  }
+
+  registerUdf(udf: Udf): void {
+    if (udf.nameParts.length <= 1) {
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/ZetaSqlWrapper.ts:144]: Looks like it is not possible to have UDF with only name without dataset. At least I don't know how to handle it)

File: server/src/bigquery/BigQueryClient.ts [hunk 3]
```diff
@@ -49,4 +64,98 @@ export class BigQueryClient implements DbtDestinationClient {
...
+      type.arrayType = {
+        elementType: BigQueryClient.toTypeProto(dataType.arrayElementType),
+      };
+    }
+    return type;
+  }
+
+  static toTypeKind(bigQueryTypeKind?: BigQueryTypeKind): TypeKind {
+    if (!bigQueryTypeKind) {
+      return TypeKind.TYPE_UNKNOWN;
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/bigquery/BigQueryClient.ts:120]: Do you want to suggest something? The code is:
```typescript
default:
  return TypeKind.TYPE_UNKNOWN;
```

File: server/src/test/ZetaSqlWrapper.spec.ts (hunk 1 above)
Comment (pgrivachev) [reply server/src/test/ZetaSqlWrapper.spec.ts:215]: Yes, added asserts.

File: server/src/ZetaSqlWrapper.ts [hunk 4]
```diff
@@ -118,6 +122,65 @@ export class ZetaSqlWrapper {
...
+      return;
+    }
+    const udfOwner = this.ensureUdfOwnerCatalogExists(udf);
+
+    if (udfOwner.customFunction?.some(c => c.namePath && arraysAreEqual(c.namePath, udf.nameParts))) {
+      return;
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/ZetaSqlWrapper.ts]: Seems like enough, let's try.

**Comment by pgrivachev** [comment]
> Some notes and 2 questions:
> 
> 1. If we can extract udf’s from AST analyse result?
> 2. Do we need to remove udf’s from catalog?

1. I will check it, probably yes.
2. We can, but I don't see why we should do it for now.

**Comment by pgrivachev** [comment]
We can't get information from AST about UDFs because if we don't register UDFs we don't have AST. Also I found one issue, fixed it and added additional tests.

**Review by fivetran-dmitryboykov — APPROVED** [review]

**Pull request merged**